File: app/src/main/java/com/tcssol/expensetracker/Data/ExpensesRepository.java
package com.tcssol.expensetracker.Data;

import android.app.Application;
import android.util.Log;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.tcssol.expensetracker.Model.Expenses;
import com.tcssol.expensetracker.Utils.ModeWrapper;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

public class ExpensesRepository {
    private final ExpenseDao expenseDao;
    private final LiveData<List<Expenses>> allExpenses;
    private final LiveData<List<Expenses>> groupedExpenses;
    private final LiveData<List<ModeWrapper>> modeDist;

    public ExpensesRepository(Application application) {
        ExpensesDatabase database = ExpensesDatabase.getDatabase(application);
        this.expenseDao = database.expenseDao();
        this.allExpenses = expenseDao.getExpenses();
        this.groupedExpenses = expenseDao.getGroupedItems();
        this.modeDist = expenseDao.getModeDist();
    }

    /** Returns all expenses synchronously for export – call from a background Executor only. */
    public List<Expenses> getAllExpensesList() {
        return expenseDao.getExpensesAllList();
    }

    public LiveData<List<Expenses>> getGroupedExpenses() {
        return groupedExpenses;
    }

    public LiveData<List<Expenses>> getAllExpenses() {
        return allExpenses;
    }

    public void insert(Expenses expenses) {
        ExpensesDatabase.databaseWriterExecutor.execute(() -> expenseDao.insertExpense(expenses));
    }

    public LiveData<Expenses> get(long id) {
        return expenseDao.get(id);
    }

    public void update(Expenses expenses) {
        ExpensesDatabase.databaseWriterExecutor.execute(() -> expenseDao.update(expenses));
    }

    public void delete(Expenses expenses) {
        ExpensesDatabase.databaseWriterExecutor.execute(() -> expenseDao.delete(expenses));
    }

    public ExpenseDao getExpenseDao() {
        return expenseDao;
    }

    public LiveData<List<Expenses>> getMonthlyExpenses(LocalDate date) {
        String year = String.valueOf(date.getYear());
        String month = String.valueOf(date.getMonthValue());
        if (month.length() == 1) month = "0" + month;
        Log.d("expd", year + "" + month);
        return expenseDao.getMonth(month, year);
    }

    public LiveData<List<Expenses>> getGroupedMonthlyExpenses(LocalDate date) {
        String year = String.valueOf(date.getYear());
        String month = String.valueOf(date.getMonthValue());
        if (month.length() == 1) month = "0" + month;
        Log.d("expd", year + "" + month);
        return expenseDao.getGroupedMonth(month, year);
    }

    /**
     * Returns MutableLiveData with [earned, spend, received, given] summaries.
     * The DAO aggregate queries must also be called on a background thread,
     * so we post the result via MutableLiveData after an Executor call.
     */
    public MutableLiveData<List<Integer>> getFrag3DataFiltered(int montht, int yearl) {
        Log.d("expdao", montht + " data received in repo " + yearl);
        String month = montht < 10 ? "0" + montht : String.valueOf(montht);
        String year = String.valueOf(yearl);
        MutableLiveData<List<Integer>> result = new MutableLiveData<>();
        ExpensesDatabase.databaseWriterExecutor.execute(() -> {
            List<Integer> list = new ArrayList<>();
            list.add(expenseDao.getSumEarnedF(month, year));
            list.add(expenseDao.getSumSpendF(month, year));
            list.add(expenseDao.getSumReceivedF(month, year));
            list.add(expenseDao.getSumGivenF(month, year));
            result.postValue(list);
        });
        return result;
    }

    public MutableLiveData<List<Integer>> getFrag3Data() {
        MutableLiveData<List<Integer>> result = new MutableLiveData<>();
        ExpensesDatabase.databaseWriterExecutor.execute(() -> {
            List<Integer> list = new ArrayList<>();
            list.add(expenseDao.getSumEarned());
            list.add(expenseDao.getSumSpend());
            list.add(expenseDao.getSumReceived());
            list.add(expenseDao.getSumGiven());
            result.postValue(list);
        });
        return result;
    }

    public List<Expenses> getSubCats(String category, boolean type) {
        // Called from Executor thread in ViewModel – safe.
        return expenseDao.getSubCats(category, type);
    }

    public List<Expenses> getSubCatsF(int montht, String year, String category, boolean type) {
        String month = montht < 10 ? "0" + montht : String.valueOf(montht);
        // Called from Executor thread in ViewModel – safe.
        return expenseDao.getSubsCatsF(month, year, category, type);
    }

    public LiveData<List<ModeWrapper>> getModeDist() {
        return modeDist;
    }

    // --- v2.0 additions ---

    public LiveData<List<Expenses>> searchExpenses(String query) {
        return expenseDao.searchExpenses(query);
    }

    public LiveData<Double> getNetBalance(String month, String year) {
        return expenseDao.getNetBalance(month, year);
    }

    /** Returns monthly spend total – call from a background Executor only. */
    public double getMonthSpend(String month, String year) {
        return expenseDao.getMonthSpend(month, year);
    }
}

File: app/src/main/java/com/tcssol/expensetracker/Data/ExpensesRepository.kt
package com.tcssol.expensetracker.Data

import android.app.Application
import com.tcssol.expensetracker.Model.DailySum
import com.tcssol.expensetracker.Model.Expenses
import com.tcssol.expensetracker.Utils.ModeWrapper
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import java.time.LocalDate

class ExpensesRepository(application: Application) {
    private val expenseDao: ExpenseDao = ExpensesDatabase.getDatabase(application).expenseDao()
    val allExpenses: Flow<List<Expenses>> = expenseDao.getExpenses()
    val groupedExpenses: Flow<List<Expenses>> = expenseDao.getGroupedItems()
    val modeDist: Flow<List<ModeWrapper>> = expenseDao.getModeDist()

    suspend fun getAllExpensesList(): List<Expenses> {
        return expenseDao.getExpensesAllList()
    }

    suspend fun insert(expenses: Expenses) {
        expenseDao.insertExpense(expenses)
    }

    fun get(id: Long): Flow<Expenses?> {
        return expenseDao.get(id)
    }

    suspend fun update(expenses: Expenses) {
        expenseDao.update(expenses)
    }

    suspend fun delete(expenses: Expenses) {
        expenseDao.delete(expenses)
    }

    fun getExpenseDao(): ExpenseDao {
        return expenseDao
    }

    fun getMonthlyExpenses(date: LocalDate): Flow<List<Expenses>> {
        val year = date.year.toString()
        var month = date.monthValue.toString()
        if (month.length == 1) month = "0$month"
        return expenseDao.getMonth(month, year)
    }

    fun getGroupedMonthlyExpenses(date: LocalDate): Flow<List<Expenses>> {
        val year = date.year.toString()
        var month = date.monthValue.toString()
        if (month.length == 1) month = "0$month"
        return expenseDao.getGroupedMonth(month, year)
    }

    fun getFrag3DataFiltered(montht: Int, yearl: Int): Flow<List<Int?>> = flow {
        val month = if (montht < 10) "0$montht" else montht.toString()
        val year = yearl.toString()
        val list = listOf(
            expenseDao.getSumEarnedF(month, year),
            expenseDao.getSumSpendF(month, year),
            expenseDao.getSumReceivedF(month, year),
            expenseDao.getSumGivenF(month, year)
        )
        emit(list)
    }

    fun getFrag3Data(): Flow<List<Int?>> = flow {
        val list = listOf(
            expenseDao.getSumEarned(),
            expenseDao.getSumSpend(),
            expenseDao.getSumReceived(),
            expenseDao.getSumGiven()
        )
        emit(list)
    }

    suspend fun getSubCats(category: String, type: Boolean): List<Expenses> {
        return expenseDao.getSubCats(category, type)
    }

    suspend fun getSubCatsF(montht: Int, year: String, category: String, type: Boolean): List<Expenses> {
        val month = if (montht < 10) "0$montht" else montht.toString()
        return expenseDao.getSubsCatsF(month, year, category, type)
    }

    fun searchExpenses(query: String): Flow<List<Expenses>> {
        return expenseDao.searchExpenses(query)
    }

    fun getNetBalance(month: String, year: String): Flow<Double?> {
        return expenseDao.getNetBalance(month, year)
    }

    fun getTotalNetBalance(): Flow<Double?> {
        return expenseDao.getTotalNetBalance()
    }

    fun getDailySums(montht: Int, yearl: Int): Flow<List<DailySum>> {
        val month = if (montht < 10) "0$montht" else montht.toString()
        val year = yearl.toString()
        return expenseDao.getDailySums(month, year)
    }

    suspend fun getMonthSpend(month: String, year: String): Double {
        return expenseDao.getMonthSpend(month, year)
    }
}
